### PR TITLE
Add LSET command

### PR DESCRIPTION
**Pull request by tisilent**
Add #272.

### CODE REVIEW

**Comment by tisilent** [comment]
@TalZaccai I will adjust the variable names later...

File: libs/server/Objects/List/ListObjectImpl.cs [hunk 1]
```diff
@@ -381,5 +381,101 @@ private void ListPop(byte* input, ref SpanByteAndMemory output, bool fDelAtHead)
...
+            var _o_curr = _out_ptr;
+            var _o_end = _o_curr + output.Length;
+
+            ObjectOutputHeader _output = default;
+
+            try
+            {
+                if (list.Count == 0)
+                {
+                    while (!RespWriteUtils.WriteError(CmdStrings.RESP_ERR_GENERIC_NOSUCHKEY.ToArray(), ref _o_curr, _o_end))
```
Comment (PaulusParssinen) [inline libs/server/Objects/List/ListObjectImpl.cs:399]:
```suggestion
                    while (!RespWriteUtils.WriteError(CmdStrings.RESP_ERR_GENERIC_NOSUCHKEY, ref _o_curr, _o_end))
```

File: libs/server/Objects/List/ListObjectImpl.cs [hunk 2]
```diff
@@ -381,5 +381,101 @@ private void ListPop(byte* input, ref SpanByteAndMemory output, bool fDelAtHead)
...
+                byte* indexParam = default;
+                var indexParamSize = 0;
+
+                // index
+                if (!RespReadUtils.ReadPtrWithLengthHeader(ref indexParam, ref indexParamSize, ref startptr, end))
+                    return;
+
+                if (NumUtils.TryBytesToInt(indexParam, indexParamSize, out var index) == false)
+                {
+                    while (!RespWriteUtils.WriteError(CmdStrings.RESP_ERR_GENERIC_VALUE_IS_NOT_INTEGER.ToArray(), ref _o_curr, _o_end))
```
Comment (PaulusParssinen) [inline libs/server/Objects/List/ListObjectImpl.cs:417]:
```suggestion
                    while (!RespWriteUtils.WriteError(CmdStrings.RESP_ERR_GENERIC_VALUE_IS_NOT_INTEGER, ref _o_curr, _o_end))
```

File: libs/server/Objects/List/ListObjectImpl.cs [hunk 3]
```diff
@@ -381,5 +381,101 @@ private void ListPop(byte* input, ref SpanByteAndMemory output, bool fDelAtHead)
...
+                        ObjectUtils.ReallocateOutput(ref output, ref isMemory, ref _out_ptr, ref ptrHandle, ref _o_curr, ref _o_end);
+                    return;
+                }
+
+                index = index < 0 ? list.Count + index : index;
+
+                if (index > list.Count - 1)
+                {
+                    while (!RespWriteUtils.WriteError(CmdStrings.RESP_ERR_GENERIC_OFFSETOUTOFRANGE.ToArray(), ref _o_curr, _o_end))
```
Comment (PaulusParssinen) [inline libs/server/Objects/List/ListObjectImpl.cs:426]:
```suggestion
                    while (!RespWriteUtils.WriteError(CmdStrings.RESP_ERR_GENERIC_OFFSETOUTOFRANGE, ref _o_curr, _o_end))
```

**Review by PaulusParssinen — CHANGES_REQUESTED** [review]
Few nits from me, we don't need to allocate the constant errors

File: libs/server/Objects/List/ListObjectImpl.cs (hunk 1 above)
Comment (tisilent) [reply libs/server/Objects/List/ListObjectImpl.cs]: I found that VS2022 often gives incorrect prompts at this position. 👻 
I'm not sure if it's caused by branch switching.

**Comment by tisilent** [comment]
@TalZaccai I have fixed the issue in NumUtil, where would it be better to add  test cases.

**Comment by TalZaccai** [comment]
> @TalZaccai I have fixed the issue in NumUtil, where would it be better to add test cases.

Good catch! Thanks!

**Review by TalZaccai — CHANGES_REQUESTED** [review]
Nice work! Added some comments.

File: libs/server/Objects/List/ListObjectImpl.cs [hunk 4]
```diff
@@ -381,5 +381,101 @@ private void ListPop(byte* input, ref SpanByteAndMemory output, bool fDelAtHead)
...
+
+                byte* _in_ptr = input + sizeof(ObjectInputHeader);
+                byte* _in_curr = _in_ptr;
+                byte* _in_end = input + length;
+
+                byte* indexParam = default;
+                var indexParamSize = 0;
+
+                // index
+                if (!RespReadUtils.ReadPtrWithLengthHeader(ref indexParam, ref indexParamSize, ref _in_curr, _in_end))
```
[the remark this answers is not in the document]
Comment (tisilent) [reply libs/server/Objects/List/ListObjectImpl.cs:412]:
```
RPUSH mylist one two three
LSET mylist a four
```
RespReadUtils.ReadString
int.TryParse

Is this feasible?

**Comment by tisilent** [comment]
@TalZaccai 
Hi.arguments error
Garnet:
`ERR wrong number of arguments for 'LSET'`
Redis:
`ERR wrong number of arguments for command`
...

**Comment by TalZaccai** [comment]
> @TalZaccai Hi.arguments error Garnet: `ERR wrong number of arguments for 'LSET'` Redis: `ERR wrong number of arguments for command` ...

I think it varies by the version of Redis you're using...
This is what I'm seeing locally:
image

File: libs/server/Objects/List/ListObjectImpl.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (tisilent) [reply libs/server/Objects/List/ListObjectImpl.cs]: @TalZaccai Hi. Can you help me check.

[the remark this answers is not in the document]
Comment (TalZaccai) [reply libs/server/Objects/List/ListObjectImpl.cs]: Not a big deal, we can leave this as-is

**Review by TalZaccai — APPROVED** [review]

File: libs/server/Objects/List/ListObjectImpl.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (tisilent) [reply libs/server/Objects/List/ListObjectImpl.cs]: Thanks, I have some lag in receiving messages... 👻

**Pull request merged**